### PR TITLE
Give TypeInfo_Class/TypeInfo_Interface.isBaseOf like C#/Java isAssignableFrom

### PR DESCRIPTION
**Pull request by n8sh**
This is equivalent to C#/Java `isAssignableFrom` ~~with the argument swapped with the receiver~~. Naming the method "isAssignableFrom" would be more familiar to people coming from C#/Java (and I would find it convenient since it's one fewer difference to remember between languages) but is potentially misleading: "alias this" and overloadable opAssign mean that this would not actually indicate whether values of one type could be assigned to another.

~~I used the name "isDerivedFrom" since the phrase "derives from" appears in the D online documentation. Swapping the argument order is not ideal, so if someone can think of a sensible name that would work with the same argument order as "isAssignableFrom" please suggest one.~~

### CODE REVIEW

File: src/object.d [hunk 1]
```diff
@@ -1736,6 +1736,42 @@ class TypeInfo_Class : TypeInfo
         }
         return o;
     }
+
+   /**
+    * Returns true if the class described by this TypeInfo_Class derives from
+    * the interface described by `parent`.
+    */
+    bool isDerivedFrom(scope const TypeInfo_Interface parent) const @nogc nothrow pure @safe
```
Comment (JinShil) [inline src/object.d]: DDoc should also include `Params` and `Returns` when applicable.

File: src/object.d [hunk 2]
```diff
@@ -1736,6 +1736,42 @@ class TypeInfo_Class : TypeInfo
...
+    bool isDerivedFrom(scope const TypeInfo_Class parent) const @nogc nothrow pure @safe
+    {
+        if (parent is null)
+            return false;
+        if (parent is this || parent.name == this.name)
+            return true;
+        return this.base !is null && this.base.isDerivedFrom(parent);
+    }
+
+    private static bool interfaceConvertibleFrom(scope const TypeInfo_Class interfaceInfo,
```
Comment (JinShil) [inline src/object.d]: I also recommend DDoc headers on private functions because it makes it easier for those browsing the code with a modern editor to see the documentation at usage sights.

**Comment by JinShil** [comment]
Seems like a good idea to me.  I'm not sure about the implementation though.  Is `name` reliable for comparisons.  I think the tests could be filled out with more corner cases (e.g. packages, nested types, multiple interface inheritance, etc.)

**Comment by Geod24** [comment]
What's the use case for this ? Java / C# reflection is primarily runtime, while D's is primarily compile time.

File: src/object.d [hunk 3]
```diff
@@ -1736,6 +1736,42 @@ class TypeInfo_Class : TypeInfo
...
+
+   /**
+    * Returns true if the class described by this TypeInfo_Class describes a class that
+    * derives from the interface described by this TypeInfo_Interface.
+    */
+    bool isDerivedFrom(scope const TypeInfo_Class parent) const @nogc nothrow pure @safe
+    {
+        if (parent is null)
+            return false;
+        if (parent is this || parent.name == this.name)
```
[the remark this answers is not in the document]
Comment (rainers) [reply src/object.d]: Dynamic casting just compares TypeInfo pointers, so if you don't expect copies of TypeInfo being made at runtime, you don't have to compare names. You might also just call `rt.cast_._d_isbaseof`.

**Comment by n8sh** [comment]
My personal use case was implementing multimethods when the classes involved aren't known at compile time.

**Comment by ZombineDev** [comment]
@n8sh Perhaps rebasing on top of master should fix the error on BuildKite.

File: src/object.d (hunk 3 above)
[the remark this answers is not in the document]
Comment (n8sh) [reply src/object.d]: If pointer comparison is good enough for dynamic casting then it is good enough for this purpose too. Thank you for the tip regarding `_d_isbaseof`.

**Comment by n8sh** [comment]
Changing to use `_d_isbaseof` and renaming to `isBaseOf` with the same argument order as `isAssignableFrom`.

File: src/object.d [hunk 4]
```diff
@@ -1102,6 +1104,24 @@ class TypeInfo_Class : TypeInfo
...
+    * Returns:
+    *  true if the class described by `child` derives from the class described
+    *  by `child`, false otherwise.
+    */
+    bool isBaseOf(scope const TypeInfo_Class child) const @nogc nothrow pure @trusted
+    {
+        for (auto ti = cast() child; ti !is null; ti = ti.base)
+            if (ti is this)
+                return true;
+        return false;
```
Comment (n8sh) [inline src/object.d]: If there's some case for classes where this gives the wrong answer but `_d_isbaseof` gives the right answer I haven't found it. Unless such a case can be found I'll leave it here since `_d_isbaseof` potentially does a lot more work and `extern(C)` function calls are opaque to the optimizer.

File: src/rt/cast_.d [hunk 5]
```diff
@@ -14,6 +14,9 @@
 module rt.cast_;
 
 extern (C):
+@nogc:
+nothrow:
+pure:
```
Comment (n8sh) [inline src/rt/cast_.d]: Adding some attributes here so the compiler can verify the attributes I claim in the `private extern(C)` declaration in `object.d`.

**Comment by thewilsonator** [comment]
@n8sh whats the status of this?

**Comment by n8sh** [comment]
@thewilsonator The code is ready but discussion is open regarding whether this should be added.

File: src/object.d [hunk 6]
```diff
@@ -1191,6 +1211,38 @@ class TypeInfo_Interface : TypeInfo
...
+    * Returns true if the class described by `child` derives from the
+    * interface described by this `TypeInfo_Interface`.
+    *
+    * Params:
+    *  child = TypeInfo for some class
+    * Returns:
+    *  true if the class described by `child` derives from the
+    *  interface described by this `TypeInfo_Interface`.
+    */
+    bool isBaseOf(scope const TypeInfo_Class child) const @nogc nothrow pure @trusted
```
Comment (n8sh) [inline src/object.d]: Maybe this and the other functions should be `final`. The limitation could be lifted later if necessary.

**Comment by thewilsonator** [comment]
It could also do with a changelog entry.

File: src/object.d (hunk 6 above)
Comment (thewilsonator) [reply src/object.d]: Sounds good to me.

File: src/object.d (hunk 4 above)
Comment (n8sh) [reply src/object.d]: Looking back at this, checking straight up the inheritance chain rather than using _d_isbaseof means that the receiver TypeInfo_Class and argument TypeInfo_Class must represent actual classes rather than being the .info field of some TypeInfo_Interface.

**Comment by thewilsonator** [comment]
Still needs a changelog entry.

**Comment by n8sh** [comment]
Any feedback regarding whether the current behavior with TypeInfo_Interface's ClassInfo is appropriate?

**Comment by thewilsonator** [comment]
None, but I don't see any reason to not have it.

**Comment by n8sh** [comment]
I'd like to hold off to see if any comments on that are forthcoming, unless you're in a hurry to move this along.

**Comment by 12345swordy** [comment]
I don't see any reason to reject this giving that there are attempts to call c# code from d.

File: src/object.d [hunk 7]
```diff
@@ -1102,6 +1104,24 @@ class TypeInfo_Class : TypeInfo
         }
         return o;
     }
+
+   /**
+    * Returns true if the class described by `child` derives from the
```
Comment (rainers) [inline src/object.d]: I think the description should explicitly mention that the class is considered a base of itself (making the name a bit misleading).

File: src/object.d [hunk 8]
```diff
@@ -1102,6 +1104,24 @@ class TypeInfo_Class : TypeInfo
...
+    * class described by this `TypeInfo_Class`.
+    *
+    * Params:
+    *  child = TypeInfo for some class
```
Comment (rainers) [inline src/object.d]: Mention that `null` is also a valid argument?

**Comment by 12345swordy** [comment]
@n8sh ping. Address rainers comments.

File: src/object.d (hunk 4 above)
Comment (n8sh) [reply src/object.d]: Solution: detect an interface's ClassInfo (TypeInfo_Class that does not represent a class) by checking if the initializer has length zero.

**Comment by 12345swordy** [comment]
@thewilsonator Any objections so far.

**Comment by thewilsonator** [comment]
None from me, but it doesn't seem to pass.

**Comment by 12345swordy** [comment]
@n8sh the test are not passing, could you rebase?

**Comment by 12345swordy** [comment]
@thewilsonator the tests have pass, need to add auto merge label to this

**Comment by Geod24** [comment]
Does this work with `shared` druntime ?

**Comment by n8sh** [comment]
@Geod24 I don't know the answer to that but it works the same as `rt.cast_._d_isbaseof`.

File: test/typeinfo/src/isbaseof.d [hunk 9]
```diff
@@ -0,0 +1,35 @@
+// https://issues.dlang.org/show_bug.cgi?id=20178
```
Comment (Geod24) [inline test/typeinfo/src/isbaseof.d]: Can you add:

- A test with a templated interface / templated class (both `C4(T) : I!T` and `C4(T) : I!int` let's say).
- A test that it can't be called with any other typeinfo (`static assert(!__traits(compiles, ...));`)

**Comment by Geod24** [comment]
Ah I found it: https://github.com/dlang/druntime/pull/2828

File: src/object.d [hunk 10]
```diff
@@ -1192,6 +1224,38 @@ class TypeInfo_Interface : TypeInfo
...
+    *
+    * Params:
+    *  child = TypeInfo for some interface
+    * Returns:
+    *  true if the interface described by `child` derives from or is
+    *  the interface described by this `TypeInfo_Interface`.
+    */
+    final bool isBaseOf(scope const TypeInfo_Interface child) const @nogc nothrow pure @trusted
+    {
+        return child !is null && cast(bool) _d_isbaseof(cast() child.info, this.info);
```
[the remark this answers is not in the document]
Comment (n8sh) [reply src/object.d]: Even better I can omit the `cast`/`!!` altogether since bool && int works.

File: src/rt/cast_.d [hunk 11]
```diff
@@ -101,7 +104,7 @@ int _d_isbaseof2(ClassInfo oc, ClassInfo c, ref size_t offset)
     return false;
 }
 
-int _d_isbaseof(ClassInfo oc, ClassInfo c)
+int _d_isbaseof(scope ClassInfo oc, scope const ClassInfo c) @safe
```
[the remark this answers is not in the document]
Comment (n8sh) [reply src/rt/cast_.d]: I'd rather not mess with the implementation of `_d_isbaseof` in this PR.

**Comment by 12345swordy** [comment]
What is the hold up here?

**Comment by 12345swordy** [comment]
The objection time limit is up
cc @thewilsonator
